src/utils/havoc.js

```diff
@@ -446,8 +446,34 @@ class ObjectValueHavocingVisitor {
   }
 
   visitAbstractValue(val: AbstractValue): void {
-    for (let i = 0, n = val.args.length; i < n; i++) {
-      this.visitValue(val.args[i]);
+    if (!val.mightBeObject()) {
+      // Only objects need to be havoced.
+      return;
+    }
+    if (val.values.isTop()) {
+      // If we don't know which object instances it might be,
+      // then it might be one of the arguments that created
+      // this value. See #2179.
+
+      // To ensure that we don't forget to provide arguments
+      // that can be havoced, we require at least one argument.
+      let whitelistedKind =
+        val.kind &&
+        (val.kind === "widened numeric property" || // TODO: Widened properties needs to be havocable.
+          val.kind.startsWith("abstractCounted"));
+      invariant(
+        whitelistedKind || val.intrinsicName || val.args.length > 0,
+        "Havoced unknown object requires havocable arguments"
+      );
+
+      for (let i = 0, n = val.args.length; i < n; i++) {
```

Comment:
Maybe leave a TODO here that havcoing all args is still overly conservative in general (e.g. for "conditional").